src/main/kotlin/com/ampnet/payoutservice/service/PayoutServiceImpl.kt

```diff
@@ -0,0 +1,70 @@
+package com.ampnet.payoutservice.service
+
+import com.ampnet.payoutservice.blockchain.BlockchainService
+import com.ampnet.payoutservice.controller.response.CreatePayoutResponse
+import com.ampnet.payoutservice.exception.ErrorCode
+import com.ampnet.payoutservice.exception.InvalidRequestException
+import com.ampnet.payoutservice.repository.MerkleTreeRepository
+import com.ampnet.payoutservice.util.BlockNumber
+import com.ampnet.payoutservice.util.ChainId
+import com.ampnet.payoutservice.util.ContractAddress
+import com.ampnet.payoutservice.util.HashFunction
+import com.ampnet.payoutservice.util.MerkleTree
+import com.ampnet.payoutservice.util.WalletAddress
+import mu.KLogging
+import org.springframework.stereotype.Service
+
+@Service
+class PayoutServiceImpl(
+    private val merkleTreeRepository: MerkleTreeRepository,
+    private val ipfsService: IpfsService,
+    private val blockchainService: BlockchainService
+) : PayoutService {
+
+    companion object : KLogging()
+
+    override fun createPayout(
+        chainId: ChainId,
+        assetAddress: ContractAddress,
+        requesterAddress: WalletAddress,
+        payoutBlock: BlockNumber
+    ): CreatePayoutResponse {
+        logger.info {
+            "Payout request for chain ID: $chainId, asset address: $assetAddress," +
+                " requester address: $requesterAddress, payout block: $payoutBlock"
+        }
+        val assetOwner = blockchainService.getAssetOwner(chainId, assetAddress)
+
+        if (assetOwner != requesterAddress) {
+            logger.warn { "Requester is not asset owner" }
+            throw InvalidRequestException(
+                ErrorCode.USER_NOT_ASSET_OWNER,
+                "User with wallet address: $requesterAddress is not the owner of asset contract: $assetAddress"
+            )
+        }
+
+        val balances = blockchainService.fetchErc20AccountBalances(
+            chainId = chainId,
+            erc20ContractAddress = assetAddress,
+            startBlock = null, // TODO provide some default values per chain ID
```

Comment:
Is it possible to get the block number when the token (contract) was deployed?

Reply:
For that we would need to fetch the contract's transaction history and get the deployment transaction... That would take even more time I believe. What we could do here is add some hard-coded map from chain ID to block number and set those block numbers to something that was at a specific date - let's say at 01-01-2020. What do you think?

Reply:
Ok, for start we set start block in properties, default value can be 0 or null.